### ISSUE TITLE
Installing WSL2 - The term 'wsl' is not recognized as the name of a cmdlet

### ISSUE THREAD

**Issue by Inari-Whitebear** [microsoft/WSL#4422]
Windows build number:  Microsoft Windows [Version 10.0.18362.295]

Trying to install WSL2 as shown on https://docs.microsoft.com/en-us/windows/wsl/wsl2-install

Executed the `Enable-WindowsOptionalFeature -Online -FeatureName VirtualMachinePlatform`, which installed and told me it needs a restart. Later after a restart, when trying the `wsl --set-version <Distro> 2` step, it goes 
```
PS C:\Users\Inari> wsl
wsl : The term 'wsl' is not recognized as the name of a cmdlet, function, script file, or operable program.
Check the spelling of the name, or if a path was included, verify that the path is correct and try again.
At line:1 char:1
+ wsl
+ ~~~
+ CategoryInfo          : ObjectNotFound: (wsl:String) [], CommandNotFoundException
+ FullyQualifiedErrorId : CommandNotFoundException
```

`where.exe wsl` or `where.exe wsl.exe` both find nothing. But I have System32 in my $Env:Path (which was the issue in another issue about this on here). `where.exe calc.exe` also finds the calc.exe in System32.

Re-running `Enable-WindowsOptionalFeature -Online -FeatureName VirtualMachinePlatform` just tells me:
```
PS C:\Users\Inari> Enable-WindowsOptionalFeature -Online -FeatureName VirtualMachinePlatform

Path          :
Online        : True
RestartNeeded : False
```
And then it still doesn't work

Running a search for wsl.exe finds 8 files:
```
C:\Windows\servicing\LCU\Package_for_RollupFix~31bf3856ad364e35~amd64~~18362.239.1.4\amd64_microsoft-windows-lxss-wsl_31bf3856ad364e35_10.0.18362.239_none_d8d55499606e362e\f\wsl.exe
C:\Windows\servicing\LCU\Package_for_RollupFix~31bf3856ad364e35~amd64~~18362.239.1.4\amd64_microsoft-windows-lxss-wsl_31bf3856ad364e35_10.0.18362.239_none_d8d55499606e362e\r\wsl.exe
C:\Windows\servicing\LCU\Package_for_RollupFix~31bf3856ad364e35~amd64~~18362.295.1.5\amd64_microsoft-windows-lxss-wsl_31bf3856ad364e35_10.0.18362.239_none_d8d55499606e362e\f\wsl.exe
C:\Windows\servicing\LCU\Package_for_RollupFix~31bf3856ad364e35~amd64~~18362.295.1.5\amd64_microsoft-windows-lxss-wsl_31bf3856ad364e35_10.0.18362.239_none_d8d55499606e362e\r\wsl.exe
C:\Windows\WinSxS\amd64_microsoft-windows-lxss-wsl_31bf3856ad364e35_10.0.18362.239_none_d8d55499606e362e\wsl.exe
C:\Windows\WinSxS\amd64_microsoft-windows-lxss-wsl_31bf3856ad364e35_10.0.18362.239_none_d8d55499606e362e\f\wsl.exe
C:\Windows\WinSxS\amd64_microsoft-windows-lxss-wsl_31bf3856ad364e35_10.0.18362.239_none_d8d55499606e362e\r\wsl.exe
C:\Windows\WinSxS\amd64_microsoft-windows-lxss-wsl.resources_31bf3856ad364e35_10.0.18362.1_en-us_1e417be3cc84e28a\wsl.exe.mui
```

**Comment by 0xbadfca11** [comment]
> Please note that you'll need to be running Windows 10 build **18917 or higher** to use WSL 2, **and that you will need to have WSL already installed** (you can find instructions to do so here).
> https://docs.microsoft.com/en-us/windows/wsl/wsl2-install

**Comment by Inari-Whitebear** [comment]
Oh, somehow I completely missed that haha. Sorry.

**Comment by tkr789** [comment]
I have done all of the above in Windows [Version 10.0.19041.172]. Reinstalled Ubuntu 18.04. In powershell:

> wsl -l:
wsl : The term 'wsl' is not recognized as the name of a cmdlet, function, script file, or operable program. Check the
spelling of the name, or if a path was included, verify that the path is correct and try again.

> wsl --set-version <Ubuntu-18.04> 2:
At line:1 char:19
wsl --set-version <Ubuntu-18.04> 2
The '<' operator is reserved for future use.
+ CategoryInfo : ParserError: (:) [], ParentContainsErrorRecordException
+ FullyQualifiedErrorId : RedirectionNotSupported

In cmd:

> wsl -l:
NAME STATE VERSION
Ubuntu-18.04 Stopped 1

wsl --set-version <Ubuntu-18.04> 2:
The system cannot find the file specified.

I have not found anything that fixes this.

**Comment by bulgariamitko** [comment]
This command actually fixed my issue:

Enable-WindowsOptionalFeature -Online -FeatureName VirtualMachinePlatform

and then I restarted and it was working

**Comment by elena-pascal** [comment]
I have the same problem:

> Image Version: 10.0.19041.388

In PowerShelll:

> $ wsl
> wsl : The term 'wsl' is not recognized as the name of a cmdlet, function, script file, or operable program. Check the
> spelling of the name, or if a path was included, verify that the path is correct and try again.
> At line:1 char:1
> + wsl
> + ~~~
>     + CategoryInfo          : ObjectNotFound: (wsl:String) [], CommandNotFoundException
>     + FullyQualifiedErrorId : CommandNotFoundException


But in Command Promp:

> wsl -l
> Windows Subsystem for Linux Distributions:
> Ubuntu-20.04 (Default)
> Ubuntu
> Ubuntu-18.04
>

**Comment by sciepsilon** [comment]
@elena-pascal, this is a known issue, although I didn't realize it at first. See the "Troubleshooting installation" section here: 
[https://docs.microsoft.com/en-us/windows/wsl/install-win10](https://docs.microsoft.com/en-us/windows/wsl/install-win10)
> The term 'wsl' is not recognized as the name of a cmdlet, function, script file, or operable program.
> 
> Ensure that the Windows Subsystem for Linux Optional Component is installed. Additionally, if you are using an ARM64 device and running this command from PowerShell, you will receive this error. Instead run wsl.exe from PowerShell Core, or Command Prompt.

PowerShell Core is an official Microsoft tool, just like normal PowerShell, but it doesn't come pre-installed. You can [download PowerShell Core from GitHub here](https://github.com/PowerShell/PowerShell/releases/tag/v6.1.3). [This article](http://techgenix.com/powershell-core/) discusses the differences between them if you're curious.

**Comment by Biswa96** [comment]
One does not need Powershell nor Powershell core nor Command Prompt to use WSL.

**Comment by sciepsilon** [comment]
@Biswa96 Most people don't - but those with ARM64 processors, including myself, do.

**Comment by sciepsilon** [comment]
@Biswa96, how can someone install and use WSL without using any of those three tools? [https://docs.microsoft.com/en-us/windows/wsl/install-win10](https://docs.microsoft.com/en-us/windows/wsl/install-win10) gives two installation methods, but both methods require running something in a terminal. Until WSL is fully installed, the only terminal options (that I'm aware of) are PowerShell, PowerShell Core, and Command Prompt. 

After installation, then yes, WSL is usable through whichever linux distro you installed, without needing to run PowerShell, PowerShell Core, or Command Prompt.

**Comment by Biswa96** [comment]
> how can someone install and use WSL without using any of those three tools?

Remember the good old days !? Try the Optional Features Program aka. `OptionalFeatures.exe`  from System32. Enable the features, restart machine, run wsl.exe.

**Comment by daledali** [comment]
Running this: 'dism.exe /online /enable-feature /featurename:Microsoft-Windows-Subsystem-Linux /all /norestart'

Fixed my problem. You need to install wsl first for it to be recognized. To be honest, the docs are quite vague but now it seems to work for me.

I got the command from https://docs.microsoft.com/en-us/windows/wsl/install-win10 in the first manual installation section

**Comment by banderlog** [comment]
> 
> 
> @elena-pascal, this is a known issue, although I didn't realize it at first. See the "Troubleshooting installation" section here:
> https://docs.microsoft.com/en-us/windows/wsl/install-win10
> 
> > The term 'wsl' is not recognized as the name of a cmdlet, function, script file, or operable program.
> > Ensure that the Windows Subsystem for Linux Optional Component is installed. Additionally, if you are using an ARM64 device and running this command from PowerShell, you will receive this error. Instead run wsl.exe from PowerShell Core, or Command Prompt.
> 
> PowerShell Core is an official Microsoft tool, just like normal PowerShell, but it doesn't come pre-installed. You can install PowerShell Core from the Microsoft Store (where it's simply called PowerShell - confusing!) or download it [from GitHub](https://github.com/PowerShell/PowerShell/releases/tag/v6.1.3). [This article](http://techgenix.com/powershell-core/) discusses the differences between them if you're curious.

can confirm that, simply running all stuff in good old `cmd` solves the issue

Also, there is a way to do ad oculus check on wsl installation status, screenshots here: https://www.windowscentral.com/install-windows-subsystem-linux-windows-10

**Comment by gabrielbino** [comment]
Também tive esse problema e não consegui resolver! Alguém?

wsl : O termo 'wsl' não é reconhecido como nome de cmdlet, função, arquivo de script ou programa operável. Verifique a
grafia do nome ou, se um caminho tiver sido incluído, veja se o caminho está correto e tente novamente.
No linha:1 caractere:1
+ wsl --set-default-version 2
+ ~~~
    + CategoryInfo          : ObjectNotFound: (wsl:String) [], CommandNotFoundException
    + FullyQualifiedErrorId : CommandNotFoundException

**Comment by gabrielbino** [comment]
> Também tive esse problema e não consegui resolver! Alguém?
> 
> wsl : O termo 'wsl' não é reconhecido como nome de cmdlet, função, arquivo de script ou programa operável. Verifique a
> grafia do nome ou, se um caminho tiver sido incluído, veja se o caminho está correto e tente novamente.
> No linha:1 caractere:1
> 
> * wsl --set-default-version 2
> * ```
>     + CategoryInfo          : ObjectNotFound: (wsl:String) [], CommandNotFoundException
>     + FullyQualifiedErrorId : CommandNotFoundException
>   ```

Solucionei o problema da seguinte maneira: 1º reiniciei o computador, por via das dúvidas. 2° utilizei o seguinte comando: "wsl -help". Em seguida executei o código que no começo tinha dado o erro acima: "wsl --set-default-version 2". Então apareceu o seguinte: 
" 
WSL 2 requer uma atualização para seu componente kernel. Para obter mais informações, visite https://aka.ms/wsl2kernel
Para obter informações sobre as principais diferenças em relação ao WSL 2, visite https://aka.ms/wsl2 
"
Um avanço já que no começo o sistema nem reconhecia o comando! Daí rodei o comando inicial novamente: "wsl --set-default-version 2" e só apareceu: 
"
Para obter informações sobre as principais diferenças em relação ao WSL 2, visite https://aka.ms/wsl2
"
Então rodei o comando: "wsl -l -v" o retorno foi: "O Subsistema do Windows para Linux não tem distribuições instaladas.
As distribuições podem ser instaladas visitando a Microsoft Store:
https://aka.ms/wslstore". Solucionei meu problema, agora vou concluir a operação seguindo o vídeo do youtube cujo link é: https://www.youtube.com/watch?v=hd6lxt5iVsg&ab_channel=Rocketseat

**Comment by banderlog** [comment]
@gabrielbino express yourself in plain English

**Comment by gabrielbino** [comment]
> @gabrielbino express yourself in plain English

Não entendi o que você quis dizer. Traduzi a frase mas continuei sem entender. De qualquer forma o problema que eu tive foi solucionado.

Abstract

I don't understand what you mean. I translated the sentence, but I still didn't understand. Anyway the problem I had was solved.

**Comment by gabrielbino** [comment]
@banderlog | I used the Google translate, sorry if there are glitches.

I solved the problem as follows: 1st restarted the computer, just in case. 2nd I used the following command: "wsl -help". Then I ran the code that at the beginning had given the above error: "wsl --set-default-version 2". Then the following appeared:
"
WSL 2 requires an update to its kernel component. For more information, visit https://aka.ms/wsl2
For information on key differences from WSL 2, visit https://aka.ms/wsl2
"
Then I ran the initial command again: "wsl --set-default-version 2"
"
I solved my problem.

**Comment by XiaoruiZhu** [comment]
For me, I need to enable two features and reboot my pc twice to make the features ready for the installation of [WSL2 Linux kernel update package for x64 machines](https://wslstorestorage.blob.core.windows.net/wslblob/wsl_update_x64.msi)

dism.exe /online /enable-feature /featurename:Microsoft-Windows-Subsystem-Linux /all /norestart

dism.exe /online /enable-feature /featurename:VirtualMachinePlatform /all /norestart

**Comment by gabrielbino** [comment]
> For me, I need to enable two features and reboot my pc twice to make the features ready for the installation of [WSL2 Linux kernel update package for x64 machines](https://wslstorestorage.blob.core.windows.net/wslblob/wsl_update_x64.msi)
> 
> `dism.exe /online /enable-feature /featurename:Microsoft-Windows-Subsystem-Linux /all /norestart`
> 
> `dism.exe /online /enable-feature /featurename:VirtualMachinePlatform /all /norestart`

Yes, correct

**Comment by ws2801** [comment]
After going through all the suggestion here, the command still does not work for me.  Then instead of running in Powershell, I try running on Command Prompt and voila, it works! 
So probably Powershell pathing may not be updated after all the steps or something like that.

Hope this helps other trying to solve it.

**Comment by bramuccci** [comment]
I had this error. The problem was that I opened "PowerShell (x86)" and not "PowerShell". I wrote the command in the later and it worked.

**Comment by cryptolodge** [comment]
My issue was that I was using x86 as well. I used PowerShell ISE instead and it works!

**Comment by yhammami** [comment]
Don't use powershell , just use cmd  :
C:\Users\YOUSSEF>wsl --install
The requested operation requires elevation.
Downloading: Windows Subsystem for Linux 2.3.26
Installing: Windows Subsystem for Linux 2.3.26
Windows Subsystem for Linux 2.3.26 has been installed.
The operation completed successfully.
Installing: Ubuntu
[==========================100.0%==================   
Installing: Ubuntu
Ubuntu has been installed.
Launching Ubuntu...
Installing, this may take a few minutes...
